src/tagmanager/tm_workspace.c

```diff
@@ -821,15 +826,20 @@ static GHashTable *get_includes(TMSourceFile *source, TMSourceFile **header)
 		gchar *hdr_name = g_path_get_basename(hdr_tag->name);
 		GPtrArray *tm_files = g_hash_table_lookup(theWorkspace->source_file_map, hdr_name);
 
-		if (tm_files && tm_files->len > 0)
+		if (tm_files)
 		{
-			TMSourceFile *hdr = tm_files->pdata[0];
-			gchar **hdr_comps = g_strsplit(hdr->short_name, ".", 2);
+			guint j;
 
-			if (*header == NULL && g_strcmp0(hdr_comps[0], src_comps[0]) == 0)
-				*header = hdr;
-			g_hash_table_add(includes, hdr);
-			g_strfreev(hdr_comps);
+			for (j = 0; j < tm_files->len; j++)
+			{
+				TMSourceFile *hdr = tm_files->pdata[j];
+				gchar **hdr_comps = g_strsplit(hdr->short_name, ".", 2);
+
+				if (g_strcmp0(hdr_comps[0], src_comps[0]) == 0)
```

Comment:
I think this can be improved.

We're called for a foo.c and loop over all includes here. We have the file name already, and then lookup the `TMSourceFile` (and then for every `TMSourceFile` we check if the file name is foo.h). 

Instead, we can already check for foo.h earlier (when looking up the `TMSourceFile`). And then we can return a copy (or ref?) of the `GPtrArray` as `header_candidates` instead of creating a new list. Something like this:

```
GPtrArray *tm_files = g_hash_table_lookup(theWorkspace->source_file_map, hdr_name);
if (g_strcmp0(hdr_name, src_comps[0]))
```

Perhaps it would make sense to separate this part out of `get_includes()` and make a distinct `g_hash_table_lookup()` for the expected header name? Hm but then it wouldn't depend on foo.c actually including foo.h (could also be seen as a good thing, not sure)

Reply:
I'm not sure if I understand completely what you suggest but it's true we can return the GPtrArray directly and avoid creation of the list - I just made a new commit doing this. Is this what you had in mind?

Note that the inner loop should still be there as we should also use all the includes of the same name similarly to the header file.

Separating extraction of header candidates to a separate function would be possible but it will probably lead to a lot of duplicated code.

Reply:
Ah, `g_ptr_array_find()` was introduced in glib 2.54 which we don't have. Wouldn't be hard to implemented, it's jut an extra complication and maybe using the list is a simpler solution then (despite some extra allocation which I think won't matter much in this case).

Reply:
I mean you can return the GPtrArray right after `g_hash_table_lookup(theWorkspace->source_file_map, hdr_name)`, before the inner loop. The GPtrArray has all headers with the same basename already and the inner loop only compares the basename once more.

The inner loop still adds to the `includes` hash table but it doesn't need to deal with the header candidates.

> Ah, g_ptr_array_find() was introduced in glib 2.54 which we don't have
All files in the GPtrArray have the same base name, so you just need to compare the basename against `->pdata[0]`

FWIW, we will probably bump GTK dependency to 3.24 soon and then we also get glib 2.54+ (which is a bit older isn't it?).

Reply:
> The inner loop still adds to the includes hash table but it doesn't need to deal with the header candidates.

Yeah, got it now, fixed in the last commit.

> FWIW, we will probably bump GTK dependency to 3.24 soon and then we also get glib 2.54+ (which is a bit older isn't it?).

That would fix the problem. 2.54 is from 2017.